docker/jenkins/publish-build.sh

```diff
@@ -176,13 +176,56 @@ else
   base64_contents=$(echo "$md_contents" | base64 --wrap=0)
 fi
 
-payload="{\"message\":\"Add $flower build $version in $build\",\"content\":\"$base64_contents\",\"sha\":\"$sha256\"}"
-echo "Sending to Github: $payload"
+githubUrl="https://api.github.com/repos/rstudio/latest-builds/contents/content/rstudio/$flower/$build/$version_stem.md"
 
-curl \
-  -X PUT \
-  -H "Accept: application/vnd.github.v3+json" \
-  -H "Authorization: token $pat" \
-  "https://api.github.com/repos/rstudio/latest-builds/contents/content/rstudio/$flower/$build/$version_stem.md" \
-  -d "$payload"
+# Sha is only required in the payload if it's an update, removing the sha field gives a clearer
+# error message
+payload="{\"message\":\"Add $flower build $version in $build\",\"content\":\"$base64_contents\"}"
+echo "Sending to Github: $payload"
 
+response=$(curl \
+   -X PUT \
+   -H "Accept: application/vnd.github.v3+json" \
+   -H "Authorization: token $pat" \
+   $githubUrl \
+   -d "$payload")
+
+echo "Github's Response: "
+echo $response
+errorMessage=$(echo $response | jq -r .message)
+
+# We get a null from jq if the message field doesn't exist. This field appears when there's been a problem with
+# our upload. We'll assume it's because the sha field is missing and this is an update, however we've
+# printed the response in the event there's some other issue
+if [[ $errorMessage != "null" ]]; then
+
+   echo "We need up perform an update, so we get the existing file's info"
+   getShaResponse=$(curl \
+      -X GET \
+      -H "Accept: application/vnd.github.v3+json" \
+      -H "Authorization: token $pat" \
+      $githubUrl \
+      -d "$payload")
+   echo $getShaResponse
+
+   updateSha=$(echo $getShaResponse | jq -r .sha)
```

Comment:
Do we know for sure that `jq` is installed on all the build containers? If not, do we need to update the Dockerfile to add it?

Reply:
Yeah, jq's included in all the `docker/jenkins` Dockerfiles

Reply:
Actually looks like I'm mistaken, the publish step doesn't happen in the container, it's on the build agent's ami, so I'm not sure

Reply:
Should we move the publish step into the container? I'm surprised to hear it wasn't already happening in the container

Reply:
It would give us a more controlled, repeatable environment, And I'm not sure what benefit publishing outside the container would give us

Reply:
I'm working on restructuring the linux jenkinsfile so I'll include that change in there. Graham is doing the windows one so I'll let him know

Reply:
Alright, Then to prevent failing builds, this is blocked until the merge of [#4185](https://github.com/rstudio/rstudio-pro/issues/4185) right?